lib/Dialect/Verif/Transforms/LowerContracts.cpp

```diff
@@ -0,0 +1,109 @@
+//===- LowerContracts.cpp - Formal Preparations --*- C++ -*----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Lower contracts into verif.formal tests.
+//
+//===----------------------------------------------------------------------===//
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "circt/Dialect/Verif/VerifPasses.h"
+
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace circt;
+
+namespace circt {
+namespace verif {
+#define GEN_PASS_DEF_LOWERCONTRACTSPASS
+#include "circt/Dialect/Verif/Passes.h.inc"
+} // namespace verif
+} // namespace circt
+
+using namespace mlir;
+using namespace verif;
+using namespace hw;
+
+namespace {
+struct LowerContractsPass
+    : verif::impl::LowerContractsPassBase<LowerContractsPass> {
+  void runOnOperation() override;
+};
+
+template <typename FROM, typename TO>
+void replaceContractOp(PatternRewriter &rewriter, Block *body) {
+  for (auto op : llvm::make_early_inc_range(body->getOps<FROM>())) {
+    auto label = op.getLabel();
+    StringAttr labelAttr;
+    if (label) {
+      labelAttr = rewriter.getStringAttr(label.value());
+    }
+    rewriter.replaceOpWithNewOp<TO>(op, op.getProperty(), op.getEnable(),
+                                    labelAttr);
+  }
+}
+
+struct HWOpRewritePattern : public OpRewritePattern<HWModuleOp> {
+  using OpRewritePattern<HWModuleOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(HWModuleOp op,
+                                PatternRewriter &rewriter) const override {
+    auto formalOp = rewriter.create<verif::FormalOp>(
+        op.getLoc(), op.getNameAttr(), rewriter.getDictionaryAttr({}));
```

Comment:
What if there's no `verif.contract` op in the module? In that case we want to just leave the module as-is, right? Because the top-level of the test is already wrapped in a `verif.formal` and just arbitrarily splitting out this module and proving it separately won't work.